File: app/src/main/java/com/example/tetrisapp/ui/fragment/GameOverFragment.java
package com.example.tetrisapp.ui.fragment;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.activity.OnBackPressedCallback;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.navigation.Navigation;
import androidx.room.rxjava3.EmptyResultSetException;

import com.example.tetrisapp.R;
import com.example.tetrisapp.data.local.dao.LeaderboardDao;
import com.example.tetrisapp.data.remote.LeaderboardService;
import com.example.tetrisapp.databinding.FragmentGameOverBinding;
import com.example.tetrisapp.model.local.entity.LeaderboardEntry;
import com.example.tetrisapp.ui.activity.MainActivity;
import com.example.tetrisapp.util.FirebaseTokenUtil;
import com.example.tetrisapp.util.LeaderboardUtil;
import com.example.tetrisapp.util.OnTouchListener;

import java.util.Date;

import javax.inject.Inject;

import dagger.hilt.android.AndroidEntryPoint;
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.schedulers.Schedulers;

@AndroidEntryPoint
public class GameOverFragment extends Fragment {
    public final static String TAG = "GameOverFragment";
    protected FragmentGameOverBinding binding;
    private GameOverFragmentArgs args;

    @Inject
    LeaderboardDao leaderboardDao;
    @Inject
    LeaderboardService leaderboardService;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        OnBackPressedCallback callback = new OnBackPressedCallback(true) {
            @Override
            public void handleOnBackPressed() {
                Navigation.findNavController(binding.getRoot()).popBackStack();
            }
        };
        requireActivity().getOnBackPressedDispatcher().addCallback(this, callback);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentGameOverBinding.inflate(inflater, container, false);
        args = GameOverFragmentArgs.fromBundle(getArguments());
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        updateUI();
        initClickListeners();
    }

    @SuppressLint("SetTextI18n")
    protected void updateUI() {
        binding.score.setText(args.getScore() + "");
        binding.level.setText(args.getLevel() + "");
        binding.lines.setText(args.getLines() + "");

        leaderboardDao.getBest()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(leaderboardEntry -> {
                    int currentHighScore = leaderboardEntry.score;

                    binding.tvHighScore.setText(currentHighScore >= args.getScore() ?
                            String.format(getString(R.string.current_high_score), currentHighScore) :
                            String.format(getString(R.string.new_high_score), args.getScore()));

                    insertScoreInDB(args.getScore(), args.getLevel(), args.getLines());
                }, throwable -> {
                    if (throwable instanceof EmptyResultSetException) {
                        binding.tvHighScore.setText(0 >= args.getScore() ?
                                String.format(getString(R.string.current_high_score), 0) :
                                String.format(getString(R.string.new_high_score), args.getScore()));

                        insertScoreInDB(args.getScore(), args.getLevel(), args.getLines());
                    } else {
                        Log.e(TAG, throwable.getLocalizedMessage());
                    }
                });
    }

    protected void insertScoreInDB(int score, int level, int lines) {
        LeaderboardEntry entry = new LeaderboardEntry();
        entry.score = score;
        entry.level = level;
        entry.lines = lines;
        entry.date = new Date();

        FirebaseTokenUtil.getFirebaseToken(token -> {
            LeaderboardUtil leaderboardUtil = new LeaderboardUtil(token, leaderboardDao, leaderboardService);
            leaderboardUtil.insert(entry);
        });
    }

    @SuppressLint("ClickableViewAccessibility")
    protected void initClickListeners() {
        binding.btnLeave.setOnTouchListener(new OnTouchListener((MainActivity) requireActivity()));
        binding.btnLeave.setOnClickListener(v -> Navigation.findNavController(binding.getRoot()).navigate(R.id.action_gameOverFragment_to_mainMenuFragment));

        binding.btnRetry.setOnTouchListener(new OnTouchListener((MainActivity) requireActivity()));
        binding.btnRetry.setOnClickListener(v -> Navigation.findNavController(binding.getRoot()).navigate(R.id.action_gameOverFragment_to_gameFragment));

        binding.btnShare.setOnTouchListener(new OnTouchListener((MainActivity) requireActivity()));
        binding.btnShare.setOnClickListener(v -> {
            Intent shareIntent = new Intent(Intent.ACTION_SEND);
            shareIntent.setType("text/plain");
            shareIntent.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.share_subject));
            shareIntent.putExtra(
                    Intent.EXTRA_TEXT,
                    String.format(getString(R.string.share_text),
                            GameOverFragmentArgs.fromBundle(getArguments()).getScore())
            );
            startActivity(Intent.createChooser(shareIntent, getString(R.string.share_using)));
        });
    }
}

File: app/src/main/java/com/example/tetrisapp/ui/fragment/GameOverFragment.kt
package com.example.tetrisapp.ui.fragment

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.activity.OnBackPressedCallback
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import androidx.navigation.Navigation.findNavController
import com.example.tetrisapp.R
import com.example.tetrisapp.data.local.dao.LeaderboardDao
import com.example.tetrisapp.databinding.FragmentGameOverBinding
import com.example.tetrisapp.model.local.LeaderboardEntry
import com.example.tetrisapp.util.FirebaseTokenUtil
import com.example.tetrisapp.util.LeaderboardUtil
import com.example.tetrisapp.util.OnTouchListener
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.util.*
import javax.inject.Inject

@AndroidEntryPoint
open class GameOverFragment : Fragment() {
    protected lateinit var binding: FragmentGameOverBinding
    private lateinit var args: GameOverFragmentArgs

    @Inject lateinit var leaderboardDao: LeaderboardDao

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val callback: OnBackPressedCallback = object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                findNavController(binding.root).popBackStack()
            }
        }
        requireActivity().onBackPressedDispatcher.addCallback(this, callback)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = FragmentGameOverBinding.inflate(inflater, container, false)
        args = GameOverFragmentArgs.fromBundle(requireArguments())
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        updateUI()
        initClickListeners()
    }

    @SuppressLint("SetTextI18n")
    protected open fun updateUI() {
        binding.score.text = args.data.score.toString()
        binding.level.text = args.data.level.toString()
        binding.lines.text = args.data.lines.toString()
        lifecycleScope.launch(Dispatchers.IO) {
            leaderboardDao.getBest().let {
                val currentHighScore = it?.score ?: 0
                binding.tvHighScore.text =
                    if (currentHighScore >= args.data.score)
                        getString(R.string.current_high_score).format(currentHighScore)
                    else
                        getString(R.string.new_high_score).format(args.data.score)

                if (args.data.score > 0) {
                    insertScoreInDB(args.data.score, args.data.level, args.data.lines, args.data.timer)
                }
            }
        }
    }

    protected fun insertScoreInDB(score: Int, level: Int, lines: Int, timeInGame: Int) {
        val entry = LeaderboardEntry()
        entry.score = score
        entry.level = level
        entry.lines = lines
        entry.date = Date()
        entry.timeInGame = timeInGame
        FirebaseTokenUtil.getFirebaseToken { token: String? ->
            val leaderboardUtil = LeaderboardUtil(token, getString(R.string.api_url), leaderboardDao)
            leaderboardUtil.insert(entry)
        }
    }

    protected open fun initClickListeners() {
        binding.btnLeave.setOnClickListener(OnTouchListener(requireActivity()) {
            findNavController(binding.root).navigate(R.id.action_gameOverFragment_to_mainMenuFragment)
        })
        binding.btnRetry.setOnClickListener(OnTouchListener(requireActivity()) {
            findNavController(binding.root).navigate(R.id.action_gameOverFragment_to_gameFragment)
        })
        binding.btnShare.setOnClickListener(OnTouchListener(requireActivity()) {
            val shareIntent = Intent(Intent.ACTION_SEND)
            shareIntent.type = "text/plain"
            shareIntent.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.share_subject))
            shareIntent.putExtra(
                Intent.EXTRA_TEXT, String.format(
                    getString(R.string.share_text),
                    args.data.score
                )
            )
            startActivity(Intent.createChooser(shareIntent, getString(R.string.share_using)))
        })
    }

    companion object {
        const val TAG = "GameOverFragment"
    }
}